{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"name":"python","version":"3.11.13","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"},"kaggle":{"accelerator":"nvidiaTeslaT4","dataSources":[{"sourceId":13657170,"sourceType":"datasetVersion","datasetId":8682723},{"sourceId":257703405,"sourceType":"kernelVersion"},{"sourceId":275039595,"sourceType":"kernelVersion"}],"dockerImageVersionId":31193,"isInternetEnabled":true,"language":"python","sourceType":"notebook","isGpuEnabled":true}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"code","source":"!pip install biopython xgboost umap-learn matplotlib seaborn scikit-learn","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-10T14:36:29.269151Z","iopub.execute_input":"2025-11-10T14:36:29.269671Z","iopub.status.idle":"2025-11-10T14:36:40.622385Z","shell.execute_reply.started":"2025-11-10T14:36:29.269648Z","shell.execute_reply":"2025-11-10T14:36:40.621485Z"}},"outputs":[{"name":"stdout","text":"Collecting biopython\n  Downloading biopython-1.86-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl.metadata (13 kB)\nRequirement already satisfied: xgboost in /usr/local/lib/python3.11/dist-packages (2.0.3)\nRequirement already satisfied: umap-learn in /usr/local/lib/python3.11/dist-packages (0.5.9.post2)\nRequirement already satisfied: matplotlib in /usr/local/lib/python3.11/dist-packages (3.7.2)\nRequirement already satisfied: seaborn in /usr/local/lib/python3.11/dist-packages (0.12.2)\nRequirement already satisfied: scikit-learn in /usr/local/lib/python3.11/dist-packages (1.2.2)\nRequirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (from biopython) (1.26.4)\nRequirement already satisfied: scipy in /usr/local/lib/python3.11/dist-packages (from xgboost) (1.15.3)\nCollecting scikit-learn\n  Downloading scikit_learn-1.7.2-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.whl.metadata (11 kB)\nRequirement already satisfied: numba>=0.51.2 in /usr/local/lib/python3.11/dist-packages (from umap-learn) (0.60.0)\nRequirement already satisfied: pynndescent>=0.5 in /usr/local/lib/python3.11/dist-packages (from umap-learn) (0.5.13)\nRequirement already satisfied: tqdm in /usr/local/lib/python3.11/dist-packages (from umap-learn) (4.67.1)\nRequirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (1.3.2)\nRequirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (0.12.1)\nRequirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (4.59.0)\nRequirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (1.4.8)\nRequirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (25.0)\nRequirement already satisfied: pillow>=6.2.0 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (11.3.0)\nRequirement already satisfied: pyparsing<3.1,>=2.3.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (3.0.9)\nRequirement already satisfied: python-dateutil>=2.7 in /usr/local/lib/python3.11/dist-packages (from matplotlib) (2.9.0.post0)\nRequirement already satisfied: pandas>=0.25 in /usr/local/lib/python3.11/dist-packages (from seaborn) (2.2.3)\nRequirement already satisfied: joblib>=1.2.0 in /usr/local/lib/python3.11/dist-packages (from scikit-learn) (1.5.2)\nRequirement already satisfied: threadpoolctl>=3.1.0 in /usr/local/lib/python3.11/dist-packages (from scikit-learn) (3.6.0)\nRequirement already satisfied: llvmlite<0.44,>=0.43.0dev0 in /usr/local/lib/python3.11/dist-packages (from numba>=0.51.2->umap-learn) (0.43.0)\nRequirement already satisfied: mkl_fft in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (1.3.8)\nRequirement already satisfied: mkl_random in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (1.2.4)\nRequirement already satisfied: mkl_umath in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (0.1.1)\nRequirement already satisfied: mkl in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (2025.3.0)\nRequirement already satisfied: tbb4py in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (2022.3.0)\nRequirement already satisfied: mkl-service in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (2.4.1)\nRequirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas>=0.25->seaborn) (2025.2)\nRequirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas>=0.25->seaborn) (2025.2)\nRequirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.7->matplotlib) (1.17.0)\nRequirement already satisfied: onemkl-license==2025.3.0 in /usr/local/lib/python3.11/dist-packages (from mkl->numpy->biopython) (2025.3.0)\nRequirement already satisfied: intel-openmp<2026,>=2024 in /usr/local/lib/python3.11/dist-packages (from mkl->numpy->biopython) (2024.2.0)\nRequirement already satisfied: tbb==2022.* in /usr/local/lib/python3.11/dist-packages (from mkl->numpy->biopython) (2022.3.0)\nRequirement already satisfied: tcmlib==1.* in /usr/local/lib/python3.11/dist-packages (from tbb==2022.*->mkl->numpy->biopython) (1.4.0)\nRequirement already satisfied: intel-cmplr-lib-rt in /usr/local/lib/python3.11/dist-packages (from mkl_umath->numpy->biopython) (2024.2.0)\nRequirement already satisfied: intel-cmplr-lib-ur==2024.2.0 in /usr/local/lib/python3.11/dist-packages (from intel-openmp<2026,>=2024->mkl->numpy->biopython) (2024.2.0)\nDownloading biopython-1.86-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl (3.2 MB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m29.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n\u001b[?25hDownloading scikit_learn-1.7.2-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.whl (9.7 MB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m9.7/9.7 MB\u001b[0m \u001b[31m109.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n\u001b[?25hInstalling collected packages: scikit-learn, biopython\n  Attempting uninstall: scikit-learn\n    Found existing installation: scikit-learn 1.2.2\n    Uninstalling scikit-learn-1.2.2:\n      Successfully uninstalled scikit-learn-1.2.2\n\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\ncategory-encoders 2.7.0 requires scikit-learn<1.6.0,>=1.0.0, but you have scikit-learn 1.7.2 which is incompatible.\ncesium 0.12.4 requires numpy<3.0,>=2.0, but you have numpy 1.26.4 which is incompatible.\nsklearn-compat 0.1.3 requires scikit-learn<1.7,>=1.2, but you have scikit-learn 1.7.2 which is incompatible.\u001b[0m\u001b[31m\n\u001b[0mSuccessfully installed biopython-1.86 scikit-learn-1.7.2\n","output_type":"stream"}],"execution_count":4},{"cell_type":"code","source":"!pip install fair-esm biopython transformers","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-10T14:36:40.624028Z","iopub.execute_input":"2025-11-10T14:36:40.624274Z","iopub.status.idle":"2025-11-10T14:36:44.245214Z","shell.execute_reply.started":"2025-11-10T14:36:40.624248Z","shell.execute_reply":"2025-11-10T14:36:44.244246Z"}},"outputs":[{"name":"stdout","text":"Collecting fair-esm\n  Downloading fair_esm-2.0.0-py3-none-any.whl.metadata (37 kB)\nRequirement already satisfied: biopython in /usr/local/lib/python3.11/dist-packages (1.86)\nRequirement already satisfied: transformers in /usr/local/lib/python3.11/dist-packages (4.53.3)\nRequirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (from biopython) (1.26.4)\nRequirement already satisfied: filelock in /usr/local/lib/python3.11/dist-packages (from transformers) (3.20.0)\nRequirement already satisfied: huggingface-hub<1.0,>=0.30.0 in /usr/local/lib/python3.11/dist-packages (from transformers) (0.36.0)\nRequirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.11/dist-packages (from transformers) (25.0)\nRequirement already satisfied: pyyaml>=5.1 in /usr/local/lib/python3.11/dist-packages (from transformers) (6.0.3)\nRequirement already satisfied: regex!=2019.12.17 in /usr/local/lib/python3.11/dist-packages (from transformers) (2025.11.3)\nRequirement already satisfied: requests in /usr/local/lib/python3.11/dist-packages (from transformers) (2.32.5)\nRequirement already satisfied: tokenizers<0.22,>=0.21 in /usr/local/lib/python3.11/dist-packages (from transformers) (0.21.2)\nRequirement already satisfied: safetensors>=0.4.3 in /usr/local/lib/python3.11/dist-packages (from transformers) (0.5.3)\nRequirement already satisfied: tqdm>=4.27 in /usr/local/lib/python3.11/dist-packages (from transformers) (4.67.1)\nRequirement already satisfied: fsspec>=2023.5.0 in /usr/local/lib/python3.11/dist-packages (from huggingface-hub<1.0,>=0.30.0->transformers) (2025.10.0)\nRequirement already satisfied: typing-extensions>=3.7.4.3 in /usr/local/lib/python3.11/dist-packages (from huggingface-hub<1.0,>=0.30.0->transformers) (4.15.0)\nRequirement already satisfied: hf-xet<2.0.0,>=1.1.3 in /usr/local/lib/python3.11/dist-packages (from huggingface-hub<1.0,>=0.30.0->transformers) (1.2.0)\nRequirement already satisfied: mkl_fft in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (1.3.8)\nRequirement already satisfied: mkl_random in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (1.2.4)\nRequirement already satisfied: mkl_umath in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (0.1.1)\nRequirement already satisfied: mkl in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (2025.3.0)\nRequirement already satisfied: tbb4py in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (2022.3.0)\nRequirement already satisfied: mkl-service in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (2.4.1)\nRequirement already satisfied: charset_normalizer<4,>=2 in /usr/local/lib/python3.11/dist-packages (from requests->transformers) (3.4.4)\nRequirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.11/dist-packages (from requests->transformers) (3.11)\nRequirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/dist-packages (from requests->transformers) (2.5.0)\nRequirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.11/dist-packages (from requests->transformers) (2025.10.5)\nRequirement already satisfied: onemkl-license==2025.3.0 in /usr/local/lib/python3.11/dist-packages (from mkl->numpy->biopython) (2025.3.0)\nRequirement already satisfied: intel-openmp<2026,>=2024 in /usr/local/lib/python3.11/dist-packages (from mkl->numpy->biopython) (2024.2.0)\nRequirement already satisfied: tbb==2022.* in /usr/local/lib/python3.11/dist-packages (from mkl->numpy->biopython) (2022.3.0)\nRequirement already satisfied: tcmlib==1.* in /usr/local/lib/python3.11/dist-packages (from tbb==2022.*->mkl->numpy->biopython) (1.4.0)\nRequirement already satisfied: intel-cmplr-lib-rt in /usr/local/lib/python3.11/dist-packages (from mkl_umath->numpy->biopython) (2024.2.0)\nRequirement already satisfied: intel-cmplr-lib-ur==2024.2.0 in /usr/local/lib/python3.11/dist-packages (from intel-openmp<2026,>=2024->mkl->numpy->biopython) (2024.2.0)\nDownloading fair_esm-2.0.0-py3-none-any.whl (93 kB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m93.1/93.1 kB\u001b[0m \u001b[31m3.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25hInstalling collected packages: fair-esm\nSuccessfully installed fair-esm-2.0.0\n","output_type":"stream"}],"execution_count":5},{"cell_type":"code","source":"import os\nimport pandas as pd\nimport numpy as np\nfrom Bio import SeqIO\n\nimport torch\nimport esm\n\nfrom sklearn.preprocessing import LabelEncoder\nfrom sklearn.cluster import KMeans\nfrom sklearn.decomposition import PCA\nfrom sklearn.ensemble import RandomForestClassifier\nfrom sklearn.model_selection import train_test_split\nfrom sklearn.metrics import accuracy_score, confusion_matrix\nimport plotly.express as px\n","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-10T14:36:44.246399Z","iopub.execute_input":"2025-11-10T14:36:44.246686Z","iopub.status.idle":"2025-11-10T14:36:51.921282Z","shell.execute_reply.started":"2025-11-10T14:36:44.246651Z","shell.execute_reply":"2025-11-10T14:36:51.920671Z"}},"outputs":[],"execution_count":6},{"cell_type":"code","source":"import pandas as pd\n\nmetadata_path = \"/kaggle/input/card-data/aro_index.tsv\"\n\ndf = pd.read_csv(metadata_path, sep=\"\\t\")\nprint(df.columns)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-10T14:36:51.923065Z","iopub.execute_input":"2025-11-10T14:36:51.923450Z","iopub.status.idle":"2025-11-10T14:36:52.001452Z","shell.execute_reply.started":"2025-11-10T14:36:51.923431Z","shell.execute_reply":"2025-11-10T14:36:52.000598Z"}},"outputs":[{"name":"stdout","text":"Index(['ARO Accession', 'CVTERM ID', 'Model Sequence ID', 'Model ID',\n       'Model Name', 'ARO Name', 'Protein Accession', 'DNA Accession',\n       'AMR Gene Family', 'Drug Class', 'Resistance Mechanism',\n       'CARD Short Name'],\n      dtype='object')\n","output_type":"stream"}],"execution_count":7},{"cell_type":"code","source":"metadata_path = \"/kaggle/input/card-data/aro_index.tsv\"\n\ndf_meta = pd.read_csv(metadata_path, sep=\"\\t\", usecols=[\n    \"ARO Accession\",\n    \"ARO Name\",\n    \"Protein Accession\",\n    \"AMR Gene Family\",\n    \"Drug Class\",\n    \"Resistance Mechanism\"\n])\n\n# Convert ARO Accession into pure numeric ID (extract digits)\ndf_meta[\"ARO Accession\"] = (\n    df_meta[\"ARO Accession\"]\n    .astype(str)\n    .str.extract(r\"(\\d+)\")\n)\n\ndf_meta.dropna(subset=[\"ARO Accession\"], inplace=True)\ndf_meta[\"ARO Accession\"] = df_meta[\"ARO Accession\"].astype(str)\n","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-10T14:36:52.002366Z","iopub.execute_input":"2025-11-10T14:36:52.002709Z","iopub.status.idle":"2025-11-10T14:36:52.058024Z","shell.execute_reply.started":"2025-11-10T14:36:52.002688Z","shell.execute_reply":"2025-11-10T14:36:52.057094Z"}},"outputs":[],"execution_count":8},{"cell_type":"code","source":"fasta_path = \"/kaggle/input/card-data/protein_fasta_protein_homolog_model.fasta\"\n\nrecords = list(SeqIO.parse(fasta_path, \"fasta\"))\n\nimport re\n\nseqs = []\n\nfor r in records:\n    # Extract ARO numeric ID from header\n    match = re.search(r\"ARO:(\\d+)\", r.description)\n    if match:\n        aro = match.group(1)\n        seqs.append({\"ARO Accession\": aro, \"sequence\": str(r.seq)})\n\ndf_seq = pd.DataFrame(seqs)\n\ndf = df_meta.merge(df_seq, on=\"ARO Accession\", how=\"inner\")\ndf.head()\nprint(df.shape)\n\n","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-10T14:36:52.058902Z","iopub.execute_input":"2025-11-10T14:36:52.059133Z","iopub.status.idle":"2025-11-10T14:36:52.258716Z","shell.execute_reply.started":"2025-11-10T14:36:52.059115Z","shell.execute_reply":"2025-11-10T14:36:52.257989Z"}},"outputs":[{"name":"stdout","text":"(6054, 7)\n","output_type":"stream"}],"execution_count":9},{"cell_type":"code","source":"from Bio import SeqIO\n\nfasta_path = \"/kaggle/input/card-data/protein_fasta_protein_homolog_model.fasta\"\n\nfor r in SeqIO.parse(fasta_path, \"fasta\"):\n    print(r.id)\n    print(r.description)\n    break\n","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-10T14:36:52.259509Z","iopub.execute_input":"2025-11-10T14:36:52.259726Z","iopub.status.idle":"2025-11-10T14:36:52.264920Z","shell.execute_reply.started":"2025-11-10T14:36:52.259711Z","shell.execute_reply":"2025-11-10T14:36:52.264275Z"}},"outputs":[{"name":"stdout","text":"gb|ACT97415.1|ARO:3002999|CblA-1\ngb|ACT97415.1|ARO:3002999|CblA-1 [mixed culture bacterium AX_gF3SD01_15]\n","output_type":"stream"}],"execution_count":10},{"cell_type":"code","source":"print(df_meta[\"ARO Accession\"].head())\nprint(df_meta[\"ARO Accession\"].dtype)\n","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-10T14:36:52.265608Z","iopub.execute_input":"2025-11-10T14:36:52.265901Z","iopub.status.idle":"2025-11-10T14:36:52.283451Z","shell.execute_reply.started":"2025-11-10T14:36:52.265870Z","shell.execute_reply":"2025-11-10T14:36:52.282859Z"}},"outputs":[{"name":"stdout","text":"0    3005099\n1    3002523\n2    3002524\n3    3002525\n4    3002526\nName: ARO Accession, dtype: object\nobject\n","output_type":"stream"}],"execution_count":11},{"cell_type":"code","source":"import re\n\nrecords = list(SeqIO.parse(fasta_path, \"fasta\"))\n\nseqs = []\n\nfor r in records:\n    desc = r.description\n\n    # Try extracting different possible ARO accession formats\n    aro_numeric = re.findall(r\"ARO:?_?(\\d+)\", desc)\n    aro_pipe = re.findall(r\"(\\d+)\\|\", desc)\n    aro_before_space = r.id.split(\".\")[0]  # part before dot\n\n    if aro_numeric:\n        aro = aro_numeric[0]\n    elif aro_pipe:\n        aro = aro_pipe[0]\n    else:\n        aro = aro_before_space\n\n    seqs.append({\"ARO Accession\": aro, \"sequence\": str(r.seq)})\n\ndf_seq = pd.DataFrame(seqs)\n\nprint(df_seq.head())\nprint(df_seq.shape)\n","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-10T14:36:52.284205Z","iopub.execute_input":"2025-11-10T14:36:52.284450Z","iopub.status.idle":"2025-11-10T14:36:52.359514Z","shell.execute_reply.started":"2025-11-10T14:36:52.284428Z","shell.execute_reply":"2025-11-10T14:36:52.358669Z"}},"outputs":[{"name":"stdout","text":"  ARO Accession                                           sequence\n0       3002999  MKAYFIAILTLFTCIATVVRAQQMSELENRIDSLLNGKKATVGIAV...\n1       3001109  MRYIRLCIISLLAALPLAVHASPQPLEQIKQSESQLSGRVGMIEMD...\n2       3002867  MIGLIVARSKNNVIGKNGNIPWKIKGEQKQFRELTTGNVVIMGRKS...\n3       3001989  MVTKRVQRMMFAAAACIPLLLGSAPLYAQTSAVQQKLAALEKSSGG...\n4       3002356  MELPNIMHPVAKLSTALAAALMLSGCMPGEIRPTIGQQMETGDQRF...\n(6052, 2)\n","output_type":"stream"}],"execution_count":12},{"cell_type":"code","source":"df = df_meta.merge(df_seq, on=\"ARO Accession\", how=\"inner\")\n\nprint(df.shape)\ndf.head()\n","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-10T14:36:52.361338Z","iopub.execute_input":"2025-11-10T14:36:52.361528Z","iopub.status.idle":"2025-11-10T14:36:52.385294Z","shell.execute_reply.started":"2025-11-10T14:36:52.361514Z","shell.execute_reply":"2025-11-10T14:36:52.384676Z"}},"outputs":[{"name":"stdout","text":"(6053, 7)\n","output_type":"stream"},{"execution_count":13,"output_type":"execute_result","data":{"text/plain":"  ARO Accession                                           ARO Name  \\\n0       3005099  23S rRNA (adenine(2058)-N(6))-methyltransferas...   \n1       3002523                                         AAC(2')-Ia   \n2       3002524                                         AAC(2')-Ib   \n3       3002525                                         AAC(2')-Ic   \n4       3002526                                         AAC(2')-Id   \n\n  Protein Accession                          AMR Gene Family  \\\n0        AAB60941.1  Erm 23S ribosomal RNA methyltransferase   \n1        AAA03550.1                                  AAC(2')   \n2        AAC44793.1                                  AAC(2')   \n3        CCP42991.1                                  AAC(2')   \n4        AAB41701.1                                  AAC(2')   \n\n                                          Drug Class  \\\n0  lincosamide antibiotic;macrolide antibiotic;st...   \n1                          aminoglycoside antibiotic   \n2                          aminoglycoside antibiotic   \n3                          aminoglycoside antibiotic   \n4                          aminoglycoside antibiotic   \n\n           Resistance Mechanism  \\\n0  antibiotic target alteration   \n1       antibiotic inactivation   \n2       antibiotic inactivation   \n3       antibiotic inactivation   \n4       antibiotic inactivation   \n\n                                            sequence  \n0  MKQKNPKNTQNFITSKKHVKEILKYTNINKQDKIIEIGSGKGHFTK...  \n1  MGIEYRSLHTSQLTLSEKEALYDLLIEGFEGDFSHDDFAHTLGGMH...  \n2  MPFQDVSAPVRGGILHTARLVHTSDLDQETREGARRMVIEAFEGDF...  \n3  MHTQVHTARLVHTADLDSETRQDIRQMVTGAFAGDFTETDWEHTLG...  \n4  MLTQHVSEARTRGAIHTARLIHTSDLDQETRDGARRMVIEAFRDPS...  ","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>ARO Accession</th>\n      <th>ARO Name</th>\n      <th>Protein Accession</th>\n      <th>AMR Gene Family</th>\n      <th>Drug Class</th>\n      <th>Resistance Mechanism</th>\n      <th>sequence</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>3005099</td>\n      <td>23S rRNA (adenine(2058)-N(6))-methyltransferas...</td>\n      <td>AAB60941.1</td>\n      <td>Erm 23S ribosomal RNA methyltransferase</td>\n      <td>lincosamide antibiotic;macrolide antibiotic;st...</td>\n      <td>antibiotic target alteration</td>\n      <td>MKQKNPKNTQNFITSKKHVKEILKYTNINKQDKIIEIGSGKGHFTK...</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>3002523</td>\n      <td>AAC(2')-Ia</td>\n      <td>AAA03550.1</td>\n      <td>AAC(2')</td>\n      <td>aminoglycoside antibiotic</td>\n      <td>antibiotic inactivation</td>\n      <td>MGIEYRSLHTSQLTLSEKEALYDLLIEGFEGDFSHDDFAHTLGGMH...</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>3002524</td>\n      <td>AAC(2')-Ib</td>\n      <td>AAC44793.1</td>\n      <td>AAC(2')</td>\n      <td>aminoglycoside antibiotic</td>\n      <td>antibiotic inactivation</td>\n      <td>MPFQDVSAPVRGGILHTARLVHTSDLDQETREGARRMVIEAFEGDF...</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>3002525</td>\n      <td>AAC(2')-Ic</td>\n      <td>CCP42991.1</td>\n      <td>AAC(2')</td>\n      <td>aminoglycoside antibiotic</td>\n      <td>antibiotic inactivation</td>\n      <td>MHTQVHTARLVHTADLDSETRQDIRQMVTGAFAGDFTETDWEHTLG...</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>3002526</td>\n      <td>AAC(2')-Id</td>\n      <td>AAB41701.1</td>\n      <td>AAC(2')</td>\n      <td>aminoglycoside antibiotic</td>\n      <td>antibiotic inactivation</td>\n      <td>MLTQHVSEARTRGAIHTARLIHTSDLDQETRDGARRMVIEAFRDPS...</td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"metadata":{}}],"execution_count":13},{"cell_type":"code","source":"device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n\nmodel, alphabet = esm.pretrained.esm2_t33_650M_UR50D()\nmodel = model.to(device)\nbatch_converter = alphabet.get_batch_converter()\n\nembeddings = []\n\nBATCH_SIZE = 16  # adjust based on GPU memory\n\nsequence_list = list(df[\"sequence\"])\n\nfor i in range(0, len(sequence_list), BATCH_SIZE):\n    batch_seqs = sequence_list[i : i + BATCH_SIZE]\n    batch = [(\"protein\", seq) for seq in batch_seqs]\n\n    labels, strs, tokens = batch_converter(batch)\n    tokens = tokens.to(device)\n\n    with torch.no_grad():\n        results = model(tokens, repr_layers=[33])\n\n    reps = results[\"representations\"][33].cpu().numpy()\n    reps_mean = reps.mean(axis=1)  # mean pool\n\n    embeddings.extend(reps_mean)\n\ndf[\"embedding\"] = embeddings\ndf.to_pickle(\"card_embeddings.pkl\")\n\nprint(\"✅ Saved embeddings to card_embeddings.pkl\")\n","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-10T09:01:01.008236Z","iopub.execute_input":"2025-11-10T09:01:01.008528Z","iopub.status.idle":"2025-11-10T09:22:24.142001Z","shell.execute_reply.started":"2025-11-10T09:01:01.008489Z","shell.execute_reply":"2025-11-10T09:22:24.141309Z"}},"outputs":[{"name":"stdout","text":"✅ Saved embeddings to card_embeddings.pkl\n","output_type":"stream"}],"execution_count":28},{"cell_type":"code","source":"df.head()\n","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-10T09:22:24.143125Z","iopub.execute_input":"2025-11-10T09:22:24.143349Z","iopub.status.idle":"2025-11-10T09:22:24.156256Z","shell.execute_reply.started":"2025-11-10T09:22:24.143332Z","shell.execute_reply":"2025-11-10T09:22:24.155716Z"}},"outputs":[{"execution_count":29,"output_type":"execute_result","data":{"text/plain":"  ARO Accession                                           ARO Name  \\\n0       3005099  23S rRNA (adenine(2058)-N(6))-methyltransferas...   \n1       3002523                                         AAC(2')-Ia   \n2       3002524                                         AAC(2')-Ib   \n3       3002525                                         AAC(2')-Ic   \n4       3002526                                         AAC(2')-Id   \n\n  Protein Accession                          AMR Gene Family  \\\n0        AAB60941.1  Erm 23S ribosomal RNA methyltransferase   \n1        AAA03550.1                                  AAC(2')   \n2        AAC44793.1                                  AAC(2')   \n3        CCP42991.1                                  AAC(2')   \n4        AAB41701.1                                  AAC(2')   \n\n                                          Drug Class  \\\n0  lincosamide antibiotic;macrolide antibiotic;st...   \n1                          aminoglycoside antibiotic   \n2                          aminoglycoside antibiotic   \n3                          aminoglycoside antibiotic   \n4                          aminoglycoside antibiotic   \n\n           Resistance Mechanism  \\\n0  antibiotic target alteration   \n1       antibiotic inactivation   \n2       antibiotic inactivation   \n3       antibiotic inactivation   \n4       antibiotic inactivation   \n\n                                            sequence  \\\n0  MKQKNPKNTQNFITSKKHVKEILKYTNINKQDKIIEIGSGKGHFTK...   \n1  MGIEYRSLHTSQLTLSEKEALYDLLIEGFEGDFSHDDFAHTLGGMH...   \n2  MPFQDVSAPVRGGILHTARLVHTSDLDQETREGARRMVIEAFEGDF...   \n3  MHTQVHTARLVHTADLDSETRQDIRQMVTGAFAGDFTETDWEHTLG...   \n4  MLTQHVSEARTRGAIHTARLIHTSDLDQETRDGARRMVIEAFRDPS...   \n\n                                           embedding  \n0  [-0.0021094177, -0.03240475, -0.0039766748, 0....  \n1  [0.0049595386, -0.07888554, 0.09245716, 0.0065...  \n2  [0.016762396, -0.048971687, 0.057476453, -0.02...  \n3  [-0.016126482, -0.06911608, 0.04928137, -0.044...  \n4  [-0.0071755834, -0.011755409, 0.05042654, -0.0...  ","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>ARO Accession</th>\n      <th>ARO Name</th>\n      <th>Protein Accession</th>\n      <th>AMR Gene Family</th>\n      <th>Drug Class</th>\n      <th>Resistance Mechanism</th>\n      <th>sequence</th>\n      <th>embedding</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>3005099</td>\n      <td>23S rRNA (adenine(2058)-N(6))-methyltransferas...</td>\n      <td>AAB60941.1</td>\n      <td>Erm 23S ribosomal RNA methyltransferase</td>\n      <td>lincosamide antibiotic;macrolide antibiotic;st...</td>\n      <td>antibiotic target alteration</td>\n      <td>MKQKNPKNTQNFITSKKHVKEILKYTNINKQDKIIEIGSGKGHFTK...</td>\n      <td>[-0.0021094177, -0.03240475, -0.0039766748, 0....</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>3002523</td>\n      <td>AAC(2')-Ia</td>\n      <td>AAA03550.1</td>\n      <td>AAC(2')</td>\n      <td>aminoglycoside antibiotic</td>\n      <td>antibiotic inactivation</td>\n      <td>MGIEYRSLHTSQLTLSEKEALYDLLIEGFEGDFSHDDFAHTLGGMH...</td>\n      <td>[0.0049595386, -0.07888554, 0.09245716, 0.0065...</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>3002524</td>\n      <td>AAC(2')-Ib</td>\n      <td>AAC44793.1</td>\n      <td>AAC(2')</td>\n      <td>aminoglycoside antibiotic</td>\n      <td>antibiotic inactivation</td>\n      <td>MPFQDVSAPVRGGILHTARLVHTSDLDQETREGARRMVIEAFEGDF...</td>\n      <td>[0.016762396, -0.048971687, 0.057476453, -0.02...</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>3002525</td>\n      <td>AAC(2')-Ic</td>\n      <td>CCP42991.1</td>\n      <td>AAC(2')</td>\n      <td>aminoglycoside antibiotic</td>\n      <td>antibiotic inactivation</td>\n      <td>MHTQVHTARLVHTADLDSETRQDIRQMVTGAFAGDFTETDWEHTLG...</td>\n      <td>[-0.016126482, -0.06911608, 0.04928137, -0.044...</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>3002526</td>\n      <td>AAC(2')-Id</td>\n      <td>AAB41701.1</td>\n      <td>AAC(2')</td>\n      <td>aminoglycoside antibiotic</td>\n      <td>antibiotic inactivation</td>\n      <td>MLTQHVSEARTRGAIHTARLIHTSDLDQETRDGARRMVIEAFRDPS...</td>\n      <td>[-0.0071755834, -0.011755409, 0.05042654, -0.0...</td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"metadata":{}}],"execution_count":29},{"cell_type":"code","source":"kmeans = KMeans(n_clusters=4, random_state=42)\ndf[\"cluster\"] = kmeans.fit_predict(embed_matrix)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-10T14:36:52.469047Z","iopub.status.idle":"2025-11-10T14:36:52.469384Z","shell.execute_reply.started":"2025-11-10T14:36:52.469273Z","shell.execute_reply":"2025-11-10T14:36:52.469284Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"pca = PCA(2)\nembed_2d = pca.fit_transform(embed_matrix)\n\ndf[\"PC1\"] = embed_2d[:,0]\ndf[\"PC2\"] = embed_2d[:,1]\n\npx.scatter(\n    df, x=\"PC1\", y=\"PC2\",\n    color=\"cluster\",\n    hover_data=[\"ARO Name\", \"Drug Class\"]\n)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-10T14:36:52.470394Z","iopub.status.idle":"2025-11-10T14:36:52.470696Z","shell.execute_reply.started":"2025-11-10T14:36:52.470548Z","shell.execute_reply":"2025-11-10T14:36:52.470562Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"df[\"target\"] = LabelEncoder().fit_transform(df[\"Model Type\"])\n\nX_train, X_test, y_train, y_test = train_test_split(\n    embed_matrix, df[\"target\"], test_size=0.2, random_state=42\n)\n\nmodel_rf = RandomForestClassifier()\nmodel_rf.fit(X_train, y_train)\n\npred = model_rf.predict(X_test)\n\nprint(\"Accuracy:\", accuracy_score(y_test, pred))\nprint(confusion_matrix(y_test, pred))\n","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-10T14:36:52.471951Z","iopub.status.idle":"2025-11-10T14:36:52.472272Z","shell.execute_reply.started":"2025-11-10T14:36:52.472093Z","shell.execute_reply":"2025-11-10T14:36:52.472108Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"df.to_csv(\"ARG_mobility_results.csv\", index=False)\nprint(\"✅ Saved all results.\")","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-10T09:32:35.197463Z","iopub.execute_input":"2025-11-10T09:32:35.197775Z","iopub.status.idle":"2025-11-10T09:32:35.810047Z","shell.execute_reply.started":"2025-11-10T09:32:35.197754Z","shell.execute_reply":"2025-11-10T09:32:35.809249Z"}},"outputs":[{"name":"stdout","text":"✅ Saved all results.\n","output_type":"stream"}],"execution_count":35}]}